### ISSUE TITLE
Flatten union within impl async_graphql::Interface and enable common fields to be defined within interface struct

### ISSUE THREAD

**Issue by chipsenkbeil** [async-graphql/async-graphql#334]
## Description of the feature

From what I've read in #29 and other issues, I don't think this is possible in the same way that flattening an interface isn't possible, right? Even if it is, I would need to be able to apply it on a method defined in an impl of **async_graphql::Object**. My current situation for [vimwiki-server](https://github.com/chipsenkbeil/vimwiki-rs/tree/master/vimwiki-server) is that I have a [Node](https://github.com/chipsenkbeil/vimwiki-rs/blob/master/vimwiki-server/src/program/graphql/elements/node.rs) that contains a variety of fields to traverse a tree alongside its data, an [Element](https://github.com/chipsenkbeil/vimwiki-rs/blob/master/vimwiki-server/src/program/graphql/elements/mod.rs#L77), which is a union of unions.

I have some types like these:

```rust
struct Node {
    // ... some data files
}

#[async_graphql::Object]
impl Node {
    async fn next_node(&self) -> Node {
        // ...
    }

    // ...

    async fn element(&self) -> Element {
        // ...
    }
}

#[async_graphql::Union]
enum Element {
    #[graphql(flatten)]
    Data1(Data),

    // ...
}
```

```graphql
{
    node {
        nextNode {
            """ some fields here related to a node """
            element {
                ... on SomeActualObject {
                    """ some fields here related to an object """
                }
                ... on AnotherObject {
                    """ some fields here related to another object """
                }
            }
        }
        element {
            ... on SomeActualObject {
                """ some fields here related to an object """
            }
            ... on AnotherObject {
                """ some fields here related to another object """
            }
        }
    }
}
```

## Code example (if possible)

Ideally, I'd like to be able to use the same flatten attribute on my returned field:

```rust
#[async_graphql::Object]
impl Node {
    async fn next_node(&self) -> Node {
        // ...
    }

    // ...

    #[graphql(flatten)]
    async fn element(&self) -> Element {
        // ...
    }
}
```

```graphql
{
    node {
        nextNode {
            """ some fields here related to a node """
            ... on SomeActualObject {
                """ some fields here related to an object """
            }
            ... on AnotherObject {
                """ some fields here related to another object """
            }
        }
        ... on SomeActualObject {
            """ some fields here related to an object """
        }
        ... on AnotherObject {
            """ some fields here related to another object """
        }
    }
}
```

**Comment by sunli829** [comment]
Sorry, I cannot add this feature because it does not comply with the GraphQL specification.

**Comment by chipsenkbeil** [comment]
@sunli829, are you sure? The way I think about it, I want my struct `Node` to be an interface whose methods are common across all of the implementors defined by the element field. This might be a duplicate of #241, now that I look at it.

[This is supported by GraphQL via interfaces](https://graphql.org/learn/schema/#interfaces), it's just that the way it's done in Rust is a little bit different.

I want to be able to have an interface where all of the common methods are on my struct, not on the individual instances, and then provide a single method that yields the specific implementations as a union. This is also something that Juniper is offering via [an enum value proc macro](https://graphql-rust.github.io/juniper/master/types/interfaces.html#enum-values-default), [trait objects](https://graphql-rust.github.io/juniper/master/types/interfaces.html#trait-object-values), or[downcasting in a trait](https://graphql-rust.github.io/juniper/master/types/interfaces.html#downcasting) (following [GraphQL syntax for downcasting interface implementer.](https://spec.graphql.org/June2018/#example-5cc55)).

From Juniper, the downcasting gets me the closest, but I'd have to write all of the casting methods by hand. If I could just flag a method as returning the enum that can be used for downcasting to an interface implementer while still being able to have the struct provide all of the common fields, that would be ideal.

**Comment by chipsenkbeil** [comment]
Also renamed the request based on what I think follows the GraphQL spec and offerings from Juniper. Seems like I would want my main struct to be an interface. The main struct would provide the common fields, rather than the implementors. And the implementors would just provide the fields for their downcasted interface implementations.

**Comment by sunli829** [comment]
I see what you mean, you want the implementation of public fields in the interface to be done only once, rather than re-implementing those fields for each object.
Since `async-graphql` implements the interface differently than `juniper`, based on function forwarding rather than `trait`, it is difficult to implement this feature. But the current implementation would be much easier to use.
I need more time to think carefully about how to solve this problem.

**Comment by chipsenkbeil** [comment]
@sunli829, I can definitely get by without this feature so no need to rush into it, but I would label it as a "nice to have." It would clean up some of the unnecessary nesting that's occurred in my server.
